src/OrchardCore/OrchardCore.DisplayManagement/Implementation/DefaultShapeFactory.cs

```diff
@@ -30,33 +31,47 @@ public class DefaultShapeFactory : DynamicObject, IShapeFactory
 
         public override bool TryInvokeMember(InvokeMemberBinder binder, object[] args, out object result)
         {
-            result = ShapeFactoryExtensions.Create(this, binder.Name, Arguments.From(args, binder.CallInfo.ArgumentNames));
+            // If the member ends with Async then return the task, otherwise await it here.
+            // Example:
+            // await New.FooAsync()
+            // New.Foo()
+
+            if (binder.Name.EndsWith("Async"))
```

Comment:
Would this be more reliable if we checked the return type as opposed to relying on people suffixing their methods with `Async` if they're async?

Something along the lines of:


    var returnType = binder.ReturnType;
    var binderReturnsTask = returnType == typeof(Task) ||
        (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(Task<>));

    if (binderReturnsTask)
    {

    }

(Totally untested)

Reply:
Someone enters this code it's when a view contains something like `New.Foo()` which creates a new shape of ShapeType "Foo" from the class `Shape`. `New` is `dynamic` so the returned type is also `dynamic`. Your code won't work.

The idea is that creating a shape is an asynchronous piece of code, but all the current usages of `New.XXX` are not awaited, so we need to await them by ourselves by calling `GetAwaiter().GetResult()`, or all views will be broken.

Then to make things better we actually let the user decide to await the result of this creation by sending back the `Task` (still a dynamic variable from a compiler's perspective). I chose to let the user define they want a Task and await it by using the `Async` suffix, but it could also be a parameter in the method call. Then in this case the use will just type `await New.FooAsync()` and we don't need to call GetAwaiter().GetResult()`, it's just clean.

Another solution is to change all the current usages of `New.Foo`, return the Task by default and add the await keyword. Looks like only the Pager uses it, plus a few other places. I think we should do it at least to state the correct defaults.